Consistent use of singleton types in manual. (#36591)
JuliaLang/julia: doc/src/manual/types.md

to write just `Vector` for the abstract type including all one-dimensional dense arrays of any
element type.

## [Singleton types](@id man-singleton-types)

Immutable composite types with no fields are called *singletons*. Formally, if 

> `a isa T && b isa T` implies `a === b`, then `T` is a singleton type. [^2]

Reviewer: I think the spirit of this formalism is great, but if I have

```julia
julia> abstract type A end

julia> struct B <: A end
```

and I don't define any other subtype of `A`, would it imply that `A` is also a singleton type?

Maybe making this perfect requires `isconcretetype(T)`?  Although if you say "for any `a` and `b`" I guess it could mean "for any `a` and `b` of any implementable type."

> should I mention that abstract types can't be singleton types, or is it obvious from the context?

So yes, I think this helps to make this definition clearer.

Author: Yes, I think we should stress that only concrete types can to be singletons.

to write just `Vector` for the abstract type including all one-dimensional dense arrays of any
element type.

## [Singleton types](@id man-singleton-types)

> Immutable composite types with no fields are called *singletons*. Formally, if 

Reviewer 2: ```suggestion
Immutable composite types with no fields are called *singletons*. Formally, if
```

> 1. `T` is an immutable composite type (ie defined with `struct`),

Reviewer 3: ```suggestion
1. `T` is an immutable composite type (i.e. defined with `struct`),
```

1. `a isa T && b isa T` implies `a === b`,

> then `T` is a singleton type.[^2] [`Base.issingletontype`](@ref) can be used to check if a

Reviewer 3: Is there an official convention on whether the footnote sign should go before or after the period? I saw both in the manual.

Author: The [Chicago Manual of Style](https://www.chicagomanualofstyle.org/qanda/data/faq/topics/Punctuation/faq0020.html) suggests that it goes after the period.

Author: Incidentally, since these questions are in general important for high-quality documentation, I started some discussion about referring a style guide in #36723.

type is a singleton type. [Abstract types](@ref man-abstract-types) cannot be singleton
types by construction.

From the definition, it follows that there can be only one instance of such types:

```jldoctest
julia> struct NoFields
       end

julia> NoFields() === NoFields()
true

julia> Base.issingletontype(NoFields)
true
```

The [`===`](@ref) function confirms that the constructed instances of `NoFields` are actually one
and the same.

Parametric types can be singleton types when the above condition holds. For example,
```jldoctest
julia> struct NoFieldsParam{T}
       end

julia> Base.issingletontype(NoFieldsParam) # can't be a singleton type ...
false

julia> NoFieldsParam{Int}() isa NoFieldsParam # ... because it has ...
true

julia> NoFieldsParam{Bool}() isa NoFieldsParam # ... multiple instances
true

julia> Base.issingletontype(NoFieldsParam{Int}) # parametrized, it is a singleton
true

julia> NoFieldsParam{Int}() === NoFieldsParam{Int}()
true
```

## [`Type{T}` type selectors](@id man-typet-type)

> For each type `T`, the `Type{T}` is an abstract parametric type whose only instance is the

Reviewer 3: "the `Type{T}` is ..." sounds weird, it reads "the type `T`", where in fact we mean "the `Type{T}` type". I would just remove "the":
```suggestion
For each type `T`, `Type{T}` is an abstract parametric type whose only instance is the
```

object `T`. Until we discuss [Parametric Methods](@ref) and [conversions](@ref
conversion-and-promotion), it is difficult to explain the utility of this construct, but in
short, it allows one to specialize function behavior on specific types as *values*. This is
useful for writing methods (especially parametric ones) whose behavior depends on a type
that is given as an explicit argument rather than implied by the type of one of its
arguments.

Since the definition is a little difficult to parse, let's look at some examples:

```jldoctest
julia> isa(Float64, Type{Float64})
true

julia> isa(Real, Type{Float64})
false

julia> isa(Real, Type{Real})
true

julia> isa(Float64, Type{Real})
false
```

> In other words, [`isa(A,Type{B})`](@ref) is true if and only if `A` and `B` are the same object

Reviewer 3: ```suggestion
In other words, [`isa(A, Type{B})`](@ref) is true if and only if `A` and `B` are the same object
```

and that object is a type.

> In particular, since parametric types are [invariant](@ref man-parametric-composite-types),

Reviewer 3: Sounds like the sentence is missing a "connection" word, e.g. I would expect "we have:" or something like that at the end, before the example.

```jldoctest
julia> struct TypeParamExample{T}
           x::T
       end

julia> TypeParamExample isa Type{TypeParamExample}
true

julia> TypeParamExample{Int} isa Type{TypeParamExample}
false

julia> TypeParamExample{Int} isa Type{TypeParamExample{Int}}
true
```

Without the parameter, `Type` is simply an abstract type which has
all type objects as its instances:

```jldoctest
julia> isa(Type{Float64}, Type)
true

julia> isa(Float64, Type)
true

julia> isa(Real, Type)
true
```

Any object that is not a type is not an instance of `Type`:

```jldoctest
julia> isa(1, Type)
false

julia> isa("foo", Type)
false
```

While `Type` is part of Julia's type hierarchy like any other abstract parametric type, it
is not commonly used outside method signatures except in some special cases. Another
important use case for `Type` is sharpening field types which would otherwise be captured

> less precisely, eg as [`DataType`](@ref man-declared-types) in the example below where the

Reviewer 3: ```suggestion
less precisely, e.g. as [`DataType`](@ref man-declared-types) in the example below where the
```
